app/controllers/training_data/code_tags_samples_controller.rb

```diff
@@ -0,0 +1,38 @@
+class TrainingData::CodeTagsSamplesController < ApplicationController
+  before_action :use_track, only: [:next]
+  before_action :use_sample, only: [:show]
+  before_action :ensure_trainer!
+
+  def index; end
+
+  def show; end
+
+  def next
+    sample = TrainingData::CodeTagsSample::RetrieveNext.(@track, params[:status])
+    return redirect_to training_data_root_path if sample.nil?
+
+    sample.lock_for_editing!(current_user)
+
+    redirect_to training_data_code_tags_sample_path(sample, status: params[:status])
```

Comment:
We need to do something like rescue and try again if `lock_for_editing` fails (maybe `redo` that 3 times (presumably getting a different sample each time) then give up?)